Add six independent egs mode

lzulauf opened the pull request:
I'm putting this up for discussion. A firmware .wav file containing the update is currently available (for the next two weeks or so) at https://github.com/lzulauf/qiemem_stages/actions/runs/13450870346
 
This adds a new mode (replacing the non-advanced stages mode) that supports six independent envelope generators. This is similar to the six envelope generators mode, but allows editing each envelope generator independently. It accomplishes this by only modifying the active envelope generator (selectable with the buttons above each slider). When switching envelope generators, the slider values are not immediately used - the user must move the slider in order to activate it.

Pressing a the button corresponding to a non-active envelope generator sets that envelope generator as the active envelope generator. Pressing a button corresponding to the active envelope generator enables all sliders. Therefore, setting all sliders and then double tapping each button will set six identical envelope generators.

In general, switching egs does not immediately reflect the physical position of sliders and pots. The user must move a slider sufficiently to enable it. Activating a slider also activates its pot and cv inputs. Slider leds will be green when that slider is enabled.

If no input cable is plugged into an envelope generator's gate input, then the previous envelope generator's gate input is used. This allows a single input to trigger multiple envelope generators.

lzulauf committed:
Add six independent egs mode

This adds a new mode (replacing the non-advanced stages mode) that
supports six independent envelope generators. This is similar to the six
envelope generators mode, but allows editing each envelope generator
independently. It accomplishes this by only modifying the active
envelope generator (selectable with the buttons above each slider). When
switching envelope generators, the slider values are not immediately
used - the user must move the slider in order to activate it.

Pressing a the button corresponding to a non-active envelope generator
sets that envelope generator as the active envelope generator. Pressing a
button corresponding to the active envelope generator enables all sliders.
Therefore, setting all sliders and then double tapping each button will
set six identical envelope generators.

In general, switching egs does not immediately reflect the physical
position of sliders and pots. The user must move a slider sufficiently to
enable it. Activating a slider also activates its pot and cv inputs.
Slider leds will be green when that slider is enabled.

If no input cable is plugged into an envelope generator's gate input, then
the previous envelope generator's gate input is used. This allows a single
input to trigger multiple envelope generators.

lzulauf commented:
I saw a comment from @qiemem indicating that they would like to replace the slow lfo mode (see https://github.com/qiemem/eurorack/issues/54#issuecomment-1959588538). This PR currently replaces the original stages mode (non-advanced) since the advanced mode supports everything in basic already. Would people prefer if the slow lfo mode were replaced instead?

qiemem commented on stages/stages.cc:
```diff
@@ -145,6 +145,149 @@ void Process(IOBuffer::Block* block, size_t size) {
+    // Check if button is pressed to switch channels or activate all sliders.
+    // Ignore switch presses if recently switched.
+    if (!activeChannelSwitchTime && ui.switches().pressed(ch)) {
+      // Once a switch is pressed, delay processing switches for 1/4 second.
+      // This is essentially a super basic debounce.
+      activeChannelSwitchTime = 1000;
```
You could also activate on release rather than press. That would allow faster double presses.

qiemem commented:
Wow, this is awesome! I had actually been tinkering with a similar mode that would give you basically six parallel Stages, but it was a usability nightmare. This strikes a really nice balance between power and usability! Code is very clean too; no concerns there.

I do know a lot of people use the basic mode because they find the additional changes in the advanced mode too complex, so I would probably prefer replacing Slow LFO mode. So I think order of modes would be:

1. Basic
2. Advanced
3. Independent EGs
4. DAHDSR
5. Harmonic
6. Alt harmonic

Keeps EG modes grouped without moving things around too much to break people's muscle memory (well, break my muscle memory anyway).

One idea: what if long press on active channel caused the channel to start cycling? Then you could use the mode for a bunch of complex EGs but still have really powerful LFOs available as well. Gate could either clock the LFO or reset; could see both being useful.

Also, any chance you could write up a brief description in the README? That could even just be what you have in comments to a designated section.

qiemem commented:
Oh I just realized it doesn't save slider/pot values between restarts. You probably will want to do that. See settings schema here: https://github.com/qiemem/eurorack/blob/bipolar/stages/settings.h#L60-L82. That's a good bit of data that you'll need though... 6 * 6 segments... maybe 8 bits for pot 8 bits for slider. So adding a `uint16_t[36]` should do it (we could cut that down for unused pots, but I think it would be nice to have the space allocated in case we want to use those for something later). Unfortunately, I don't remember how much space is left on flash. Also, Stages writes its settings on successive blocks to avoid writing too much to any single location and wearing out flash. So this will shorten the lifespan of the flash memory, but I don't know how much... Any ideas?

lzulauf committed:
Update available stages

Re-adds basic stages mode, removes slow lfo mode, and restores the original stage ordering.

qiemem commented on stages/stages.cc:
```diff
@@ -145,6 +145,149 @@ void Process(IOBuffer::Block* block, size_t size) {
+  // n.b. cv inputs only apply to active envelope and only if overriding dahdsr.
+  // This allows cv to directly offset the slider position. Otherwise we would need to track separate
+  // "virtual" slider positions for each envelope.
+  if (slider_enabled[0]) {
+    eg[active_envelope].SetDelayLength(block->cv_slider[0] + block->cv[0]);
+  }
+  if (slider_enabled[1]) {
+    eg[active_envelope].SetAttackLength(block->cv_slider[1] + block->cv[1]);
+    eg[active_envelope].SetAttackCurve (block->pot[1]);
+  }
+  if (slider_enabled[2]) {
+    eg[active_envelope].SetHoldLength(block->cv_slider[2] + block->cv[2]);
+  }
+  if (slider_enabled[3]) {
+    eg[active_envelope].SetDecayLength(block->cv_slider[3] + block->cv[3]);
+    eg[active_envelope].SetDecayCurve  (block->pot[3]);
+  }
+  if (slider_enabled[4]) {
+    eg[active_envelope].SetSustainLevel(block->cv_slider[4] + block->cv[4]);
+  }
+  if (slider_enabled[5]) {
+    eg[active_envelope].SetReleaseLength(block->cv_slider[5] + block->cv[5]);
+    eg[active_envelope].SetReleaseCurve(block->pot[5]);
+  }
```
Oh sorry should have caught this earlier. `block->cv_slider` is already the sum of `block->cv` and `block->slider` (constrained to [-1, 2)). So I think you just want `block->cv_slider` here. Alternatively, if you end up wanting use CV for something else, just `block->slider`.

qiemem commented:
Oh, one critical thing if you do start working on the save thing: make sure that it saves only on discrete, occasional events. For instance, segment generator modes save on [state changes that are triggered by button releases](https://github.com/qiemem/eurorack/blob/bipolar/stages/chain_state.cc#L664) so they can't occur in rapid succession. So I would either do it on button release or on, say, a 30 second debounce after slider moves (with some hysteresis so it doesn't keep triggering). If it saves too much it *will* burn out flash memory. It can be tempting to save on button down but even a brief press in that case could damage the chip.

Also, one other idea: I found the CV only affecting the active envelope to be slightly awkward. What if CV acts as a scalar for the total duration for the corresponding channel? We would definitely not want it to impact segments of 0 duration. Exactly how it impacts other segments might take some tuning to get it to feel right. Anyway, just a thought.

To be clear, happy to merge this after the mode numbers have been changed (which you've already done; thank you!) and the CV slider thing I just commented on is fixed. Docs and saving would be nice, but I can merge first if you prefer. Looping and changing CV schema are just ideas: if they sound interesting, go for it, and if not, no worries. I may give them a shot at a later date too.

lzulauf commented:
In the interest of getting this merged sooner rather than later, but also trying to avoid adding features we want to change/remove later, maybe we remove all cv input support for the time being? Then we can decide what to do with cvs in a followup PR.

Some options:
1. CVs sum with slider, but only for the active envelope (what this PR does at the time of writing this comment).
2. CVs scale entire envelope (ignoring sustain stage) - each envelope has it's own cv input. How do we map voltages to scalings? Is the multiplier just the number of volts (i.e. you would need to send 1 volt in order to _not_ affect scale? Or do we do something logarithmic such that 0 = no change, -1 = half speed, +1 = double speed?
3. Cvs sum with slider positions for each envelope generator (i.e. we store what the slider position was for each envelope and for each stage such that we can add the current cv value). For example, adding to the attack phase would increase the attack time (linearly?) to _all_ envelopes.

lzulauf committed:
Update README

Add information about 6 independent envelopes mode, remove information about slow lfo mode.

lzulauf committed:
Use slider instead of cv_slider

It was double counting cvs since they were already included in cv_slider.

Additionally, removing cv support while we decide how best to utilize the available patch inputs (see discussion in PR).

lzulauf commented:
I posted changes to the README and fix for the cv_slider usage. Let me know if you want any additional changes on this PR. I'll try to explore some of the other ideas in separate PRs:
- Switching press detection to from checking whether the button is currently pressed to instead detecting button up events
- Add save/load of the envelope states
- Add cv input support
- long-press to put envelope in cycling mode

qiemem commented:
Sounds good! Merged!

bowlneudel commented:
Thanks for this awesome new mode! I played around with it a bit yesterday, and it's amazing to have 6 fully independent envelopes coming out of my beloved Stages.

I see there is some things to be figured out with regard to state storage, and I have no business making any suggestions there. That said, I have a few notes from a UX perspective (which I hope don't complicate the flash memory issue):

- It's easy to lose the active channel indicator (Shape/Time LED) when the envelopes have medium to long release times. I suggest giving the slider LED of the active channel a slow sine pulse. Duty cycle of that pulse could drop if that slider value = 0

- I'd love to retain the coarse envelope duration ranges from the classic stages modes, using the press + slide action. This could be represented by the slider LED pulse rate on the active channel.

- The double-click to learn all slider values is very nice. Another copy/paste function could be: push source, wiggle destination (or push-hold-push)

-  My vote for CV would be full channel scaling using v/oct (doubling / halving, unless that proves too extreme). Personally, I'd find that more useful that CV over one stage across all envelopes. I wouldn't want the effect of CV to change as I switch active channels.

lzulauf commented:
@bowlneudel thank you for the feedback! Those all sound like great ideas! For copy/paste what if it were push (and hold) source and then press destination (while still holding source)? I think this would be in addition to still supporting the existing tapping the selected envelope to enable all sliders (but would allow copying envelopes regardless of slider positions).

bowlneudel commented:
Thank YOU! Yep, that's exactly what I meant by "push-hold-push". And I agree: it should supplement, not replace the current "tap selected" action (although this implies button actions happen only on release)

To recycle the push + wiggle gesture, I could imagine this as the workflow behind a super complex CV map. Like, for the active channel: push to select the CV source, hold, and then wiggle the pot for the destination stage — to link and simultaneously set an attenuversion value. Is that beyond the limits of this hardware? I think it'd be very tricky to clearly display the current state of the CV map. Best I can think of right now is to have a blinky green shape/time LED = positive CV on this stage of the active channel, blinky red = negative CV, to say nothing of the source.

qiemem commented:
> The double-click to learn all slider values is very nice. Another copy/paste function could be: push source, wiggle destination (or push-hold-push)

Love the push-hold-push to copy/paste idea. Super intuitive and useful.

> My vote for CV would be full channel scaling using v/oct (doubling / halving, unless that proves too extreme). Personally, I'd find that more useful that CV over one stage across all envelopes. I wouldn't want the effect of CV to change as I switch active channels.

Ya, I think this would work really will. Some math notes here. Envelope times are an exponential-ish function of their inputs (via [WarpPhase](https://github.com/qiemem/eurorack/blob/1d401208eb011ba0c255e4da8dd3780aa3c9896b/stages/envelope.cc#L159-L170) with a shape of 0.25 and a max of 10s). So, as is, adding the CV value will act sort of like exponential scaling in the way you describe. 

That said, I'm not crazy about the response of the current stage times. Green segments in the segment generator modes use [RateToFrequency](https://github.com/qiemem/eurorack/blob/1d401208eb011ba0c255e4da8dd3780aa3c9896b/stages/segment_generator.cc#L153-L157) to get a range from 1ms to 16s. It's not exponential either (it's an 8th order polynomial) but it's closer. Here's the comparison between the curves:

![image](https://github.com/user-attachments/assets/b0ac851b-051b-41ab-a46e-173029b73541)

I like that they have a finer response at the lower voltages while getting a wider range.

Regardless, with these curves, just adding CV values might feel close enough to exponential scaling to work, but we still wouldn't want to add it to the ones that have sliders at 0. Alternatively, we could just use v/oct scaling and see how that feels.

> Add save/load of the envelope states

I added #64 to track saving/loading of the feature. Went back through the storage code to figure out what the expected impact on memory will be.

> To recycle the push + wiggle gesture, I could imagine this as the workflow behind a super complex CV map. Like, for the active channel: push to select the CV source, hold, and then wiggle the pot for the destination stage — to link and simultaneously set an attenuversion value. Is that beyond the limits of this hardware? I think it'd be very tricky to clearly display the current state of the CV map. Best I can think of right now is to have a blinky green shape/time LED = positive CV on this stage of the active channel, blinky red = negative CV, to say nothing of the source.

I've definitely been jealous of Quadrax's internal CV routing and thought about similar stuff for Stages before... It wouldn't be beyond hardware but it would take a lot more memory to save the state. As you mentioned, UI would also be a problem. Showing LED color indications while you're adjusting a particular routing might work. Anyway, it would be very complex to use and a lot to keep track of... At some point, it's probably just worth getting a matrix mixer or whatever...
